internal/biz/hosts/hosts.go

```diff
@@ -32,6 +36,19 @@ func New(repo HostRepo, logger log.Logger) *HostUsecase {
 
 // CreateHost creates a Host in the repository and returns the new Host.
 func (uc *HostUsecase) CreateHost(ctx context.Context, h *Host) (*Host, error) {
+	resourceId := common.ResourceId{
+		LocalResourceId: h.Metadata.Reporters[0].LocalResourceId,
+		ReporterType:    h.Metadata.Reporters[0].ReporterType,
+		ReporterId:      h.Metadata.Reporters[0].ReporterID,
+	}
+
+	_, err := uc.repo.FindByID(ctx, resourceId)
+	if err == nil {
+		return nil, fmt.Errorf("rhel_host with local_resource_id: `%v` already exists for current reporter", resourceId.LocalResourceId)
+	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, err
+	}
```

Comment:
Is there also a unique index defined for resource id, reporter type, reporter id?

Otherwise this will be prone to race conditions and you could still end up with duplicate resources.

Reply:
Yes, there is one at the Reporter level

https://github.com/project-kessel/inventory-api/blob/main/internal/biz/common/common.go#L34-L63

Although we are missing one for the `metadata_id` to allow the reporter to share the same `resource_id` among different types of resources.

https://github.com/project-kessel/inventory-api/pull/171